packages/terra-paginator/src/ProgressivePaginator.jsx

```diff
@@ -0,0 +1,95 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames/bind';
+import IconNext from 'terra-icon/lib/icon/IconNext';
+import IconPrevious from 'terra-icon/lib/icon/IconPrevious';
+import NumberField from 'terra-form/lib/NumberField';
+
+import 'terra-base/lib/baseStyles';
+import styles from './Paginator.scss';
+
+import { calculatePages } from './_paginationUtils';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * Function to be executed when a navigation element is selected.
+   */
+  onPageChange: PropTypes.func.isRequired,
+  /**
+   * The active/selected page.
+   */
+  selectedPage: PropTypes.number.isRequired,
+  /**
+   * Total number of all items being paginated.
+   */
+  totalCount: PropTypes.number.isRequired,
+  /**
+   * Total number of items per page.
+   */
+  itemCountPerPage: PropTypes.number.isRequired,
+};
+
+class ProgressivePaginator extends React.Component {
+  constructor(props) {
+    super(props);
+
+    const { selectedPage, totalCount } = this.props;
+
+    this.state = {
+      pageInput: selectedPage,
+      selectedPage: selectedPage && totalCount ? selectedPage : undefined,
+    };
+
+    this.handlePageChange = this.handlePageChange.bind(this);
+    this.handlePageInput = this.handlePageInput.bind(this);
+    this.buildPageSelectorForm = this.buildPageSelectorForm.bind(this);
+  }
+
+  handlePageInput(e) {
+    this.setState({ pageInput: e.target.value });
```

Comment:
Should this handle onPageChange as well?

Reply:
Actually, this should stay how it is. This function makes sure that the number entered in the form field is updated in the state correctly and then the change gets handled onSubmit. There is probably a better way to do that and I'm open to suggestions.